### PR TITLE
Tree model sort

### PR DESCRIPTION
**Pull request by mfulz**
#503 Should be done afaik.
At least the TreeModelSort is working now :)

### CODE REVIEW

File: gtk/gtk.go [hunk 1]
```diff
@@ -9065,6 +9066,21 @@ func (v *TreeModelFilter) SetVisibleColumn(column int) {
...
+		return nil
+	}
+	p := &TreePath{path}
+	return p
+}
+
+// ConvertChildIterToIter is a wrapper around gtk_tree_model_filter_convert_child_iter_to_iter().
+func (v *TreeModelFilter) ConvertChildIterToIter(sortIter, childIter *TreeIter) bool {
+	return gobool(C.gtk_tree_model_filter_convert_child_iter_to_iter(v.native(), sortIter.native(), childIter.native()))
+}
```
Comment (MJacred) [inline gtk/gtk.go]: ConvertChildIterToIter() should have only one input parameter and two return values (in the spirit of golang).

Furthermore, sortIter (please rename it to `filterIter`) may be nil according to [documentation](https://developer.gnome.org/gtk3/unstable/GtkTreeModelFilter.html#gtk-tree-model-filter-convert-child-iter-to-iter) and will therefore crash when you call it

File: gtk/gtk.go [hunk 2]
```diff
@@ -9065,6 +9066,21 @@ func (v *TreeModelFilter) SetVisibleColumn(column int) {
...
 
+// ConvertChildPathToPath is a wrapper around gtk_tree_model_filter_convert_child_path_to_path().
+func (v *TreeModelFilter) ConvertChildPathToPath(childPath *TreePath) *TreePath {
+	path := C.gtk_tree_model_filter_convert_child_path_to_path(v.native(), childPath.native())
+	if path == nil {
+		return nil
+	}
+	p := &TreePath{path}
+	return p
+}
```
Comment (MJacred) [inline gtk/gtk.go]: Please check if `childPath` is nil and do an early bail with a second return value of type `error`

also, please use the global standard error for unexpected cgo nil value called `nilPtrErr` and don't return only nil value

-> see `func ButtonNew() (*Button, error)` as reference

additionally:  
this is purely up to you, but maybe you want to write a `wrapTreePath` func (see existing `wrapTreeModel` as reference)

File: gtk/gtk.go [hunk 3]
```diff
@@ -9506,6 +9522,130 @@ func (v *TreeRowReference) Valid() bool {
...
+
+// ConvertChildIterToIter is a wrapper around gtk_tree_model_sort_convert_child_iter_to_iter().
+func (v *TreeModelSort) ConvertChildIterToIter(sortIter, childIter *TreeIter) bool {
+	return gobool(C.gtk_tree_model_sort_convert_child_iter_to_iter(v.native(), sortIter.native(), childIter.native()))
+}
```
Comment (MJacred) [inline gtk/gtk.go]: same as above:

ConvertChildIterToIter() should have only one input parameter and two return values (in the spirit of golang).

Furthermore, sortIter may be nil according to [documentation](https://developer.gnome.org/gtk3/unstable/GtkTreeModelFilter.html#gtk-tree-model-filter-convert-child-iter-to-iter) and will therefore crash when you call it

see `func (v *TreeModelFilter) ConvertIterToChildIter(filterIter *TreeIter)`

File: gtk/gtk.go [hunk 4]
```diff
@@ -9506,6 +9522,130 @@ func (v *TreeRowReference) Valid() bool {
...
+
+// ConvertPathToChildPath is a wrapper around gtk_tree_model_sort_convert_path_to_child_path().
+func (v *TreeModelSort) ConvertPathToChildPath(sortPath *TreePath) *TreePath {
+	path := C.gtk_tree_model_sort_convert_path_to_child_path(v.native(), sortPath.native())
+	p := &TreePath{path}
+	return p
+}
```
Comment (MJacred) [inline gtk/gtk.go]: the return value of gtk_tree_model_sort_convert_path_to_child_path() may be NULL

please see `func (v *TreeModelFilter) ConvertPathToChildPath(filterPath *TreePath)` as reference for fixing

File: gtk/gtk.go [hunk 5]
```diff
@@ -9506,6 +9522,130 @@ func (v *TreeRowReference) Valid() bool {
...
+
+// ConvertIterToChildIter is a wrapper around gtk_tree_model_sort_convert_iter_to_child_iter().
+func (v *TreeModelSort) ConvertIterToChildIter(childIter, sortIter *TreeIter) {
+	C.gtk_tree_model_sort_convert_iter_to_child_iter(v.native(), childIter.native(), sortIter.native())
+}
```
Comment (MJacred) [inline gtk/gtk.go]: same as above:

ConvertIterToChildIter() should have only one input parameter and two return values (in the spirit of golang).

Furthermore, childIter may be nil (even if its [documentation](https://developer.gnome.org/gtk3/unstable/GtkTreeModelSort.html#gtk-tree-model-sort-convert-iter-to-child-iter) does not state it directly: `child_iter An uninitialized GtkTreeIter.`) and will therefore crash when you call it

see `func (v *TreeModelFilter) ConvertIterToChildIter(filterIter *TreeIter)`

**Comment by MJacred** [comment]
and generally, please check if return value is nil and then put the return value into `glib.Take(unsafe.Pointer(c))`
* before storing it into a new pointer to a new struct, or
* when calling a `wrap*` function

examples:

```golang
// AssistantNew is a wrapper around gtk_assistant_new().
func AssistantNew() (*Assistant, error) {
	c := C.gtk_assistant_new()
	if c == nil {
		return nil, nilPtrErr
	}
	obj := glib.Take(unsafe.Pointer(c))
	return wrapAssistant(obj), nil
}
```

```golang
// BuilderNew is a wrapper around gtk_builder_new().
func BuilderNew() (*Builder, error) {
	c := C.gtk_builder_new()
	if c == nil {
		return nil, nilPtrErr
	}
	obj := glib.Take(unsafe.Pointer(c))
	return &Builder{obj}, nil
}
```

File: gtk/gtk.go (hunk 1 above)
Comment (founderio) [reply gtk/gtk.go]: Calls to nil are OK with golang in this case.
See definition of [TreeIter.native()](https://github.com/gotk3/gotk3/blob/69caac910e503eb4c7d2de1c82f36de61dca361a/gtk/gtk.go#L8774).
However, this should never be a nil value - just an uninitialized iter, meaning a valid pointer, just default content. (`new(gtk.TreeIter)`)

filterIter is an out variable so one could indeed argue that it should be a return value instead.

File: gtk/gtk.go (hunk 3 above)
Comment (founderio) [reply gtk/gtk.go]: See above comment for nil values in TreeIter, they are fine but using nil will not make sense and GTK will complain.

File: gtk/gtk.go (hunk 5 above)
Comment (founderio) [reply gtk/gtk.go]: See above comment for nil values in TreeIter, they are fine but using nil will not make sense and GTK will complain.

**Comment by mfulz** [comment]
I hope I've got every point. If not please tell me what's still not ok :)

**Comment by founderio** [comment]
Looks good to me, unless the travis pipeline tells you otherwise ;)

**Comment by MJacred** [comment]
looks good to me as well  
still not 100% sure when to use runtime.SetFinalizer(), though...

**Comment by founderio** [comment]
> still not 100% sure when to use runtime.SetFinalizer()

When there is something to free/unref/cleanup.
Any of these criteria:
* it is a `glib.Object` of some sort
* it is a C-style array (e.g. `char *`)
* it has a custom `free()` function (e.g. `TreePath` or `glib.List`)

TreeIter is just a plain struct, and similarly a lot of types like GdkWindowEdge are enums, i.e. integers - these don't need a finalizer. Those are taken care of automatically, should they ever escape into the heap memory.

File: gtk/gtk.go (hunk 5 above)
Comment (mfulz) [reply gtk/gtk.go]: > > still not 100% sure when to use runtime.SetFinalizer()
> 
> When there is something to free/unref/cleanup.
> Any of these criteria:
> 
>     * it is a `glib.Object` of some sort
> 
>     * it is a C-style array (e.g. `char *`)
> 
>     * it has a custom `free()` function (e.g. `TreePath` or `glib.List`)
> 
> 
> TreeIter is just a plain struct, and similarly a lot of types like GdkWindowEdge are enums, i.e. integers - these don't need a finalizer. Those are taken care of automatically, should they ever escape into the heap memory.

Thanks for the explanation, now this makes total sense to me :)

**Comment by MJacred** [comment]
@mfulz: the discussions should be resolved, no? if so, can you please mark them as such?

then @andre-hub can remove the labels and do a merge

**Comment by andre-hub** [comment]
i merged this now and all open questions and justified objections should be done on the common code base. there are some good mappings here, it would be a pity if they are not available to everyone.

**Pull request merged**